docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,294 @@
+* The new bears always have to constructed with ``section`` and ``file_dict``
+  as parameters, this is important as it makes the new dependency system
+  possible.
+* Bear writers need to keep this in mind while writing their ``__init__``
+  methods, default parameters are allowed as many as wanted but the same isn't
+  true for positional parameters (default parameters are discouraged, since the
+  dependency system automatically instantiates bears, which might lead to
+  unwanted side-effects).
+
+Official support for virtual files
+----------------------------------
+
+IDEs like IntelliJ use a virtual files to represent files in a filesystem (VFS)
+and perform operations on them. Hence NextGen-Core provides official support
+for virtual files, the bears have to point to the right file data objects when
+run, whether they are real files or virtual ones. This makes coala easier to
+integrate with IDEs since the contents reside in the IDE memory which is faster
+instead of reloading files from the disk.
+
+Task Objects
+------------
+
+These are the representation of the tasks performed by the bears. Structure-wise
+they are a tuple containing the tuple of positional arguments and a dict of
+keyword arguments to the bears. The task objects are instrumental to the caching
+mechanism as their hash values are stored in the cache along with the bear
+results and during each coala run the core looks for identical task objects in
+the cache, and fetches their corresponding results if found without even
+executing those tasks thereby providing a performance boost. Below you can see
+an example of two task objects.
+
+::
+
+    [
+        [(file, filename), {'filename_prefix': "", 'filename_suffix': ""}],
+        [(filename, file), {'preferred_quotation': ""}],
+    ]
+
+Improved dependency system
+--------------------------
+
+The NextGen-Core also introduces a better dependency management system than the
+one used by the old core. It features the following improvements:
+
+* A class ``DependencyTracker`` manages this task. The dependencies
+  can be added and resolved by this class and it also checks for circular
+  dependencies.
+* The dependency relations between two objects are tracked using a directed
+  graph. When two objects are added to this graph as an edge they form a
+  dependency relation.
+* Bears don't have to pass even results anymore.
+
+DependencyBear
+--------------
+
+The handling of bear dependencies by the old core wasn't effective. The
+old core used a queuing mechanism to communicate between bear runs which
+led to problems of a GlobalBear not running if any one of its
+dependencies (a local bear) terminated, the NextGen-Core improves on this.
+
+A new bear known as the DependencyBear is also provided by the NextGen-Core to
+make it more convenient for bear developers to write dependency bears, by
+passing the dependency results using task objects (this also provides a way
+to do efficient caching which is explained later).
```

Comment:
it even enables proper caching :)
You could mention that dependency results lie inside `self.dependency_results` inside a bear and can be accessed that way, however this is highly discouraged because this bypasses caching and could yield weird results when the core is run multiple times in a row.